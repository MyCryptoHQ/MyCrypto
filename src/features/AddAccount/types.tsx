// @ADD_ACCOUNT_@todo: move to named enum or other Set
export enum FormDataActionType {
  SELECT_NETWORK,
  SELECT_ACCOUNT,
  SELECT_ACCOUNT_TYPE,
  SET_LABEL,
  SET_DERIVATION_PATH,
  ON_UNLOCK,
  RESET_FORM
}

export interface FormDataAction {
  type: FormDataActionType;
  payload: any;
}
