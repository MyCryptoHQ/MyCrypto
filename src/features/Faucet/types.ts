import { ITxData, ITxGasLimit, ITxGasPrice, ITxHash, TAddress } from '@types';

export interface ITxFaucetResult {
  chainId: number;
  data: ITxData;
  from: TAddress;
  gasLimit: ITxGasLimit;
  gasPrice: ITxGasPrice;
  hash: ITxHash;
  network: string;
  nonce: number;
  to: TAddress;
  value: string;
}
