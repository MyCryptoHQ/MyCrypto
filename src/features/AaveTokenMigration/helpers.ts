import { DEFAULT_ASSET_DECIMAL, DEFAULT_NETWORK_CHAINID } from '@config';
import { formatApproveTx } from '@helpers';
import { AaveMigrator } from '@services/EthService/contracts';
import { ITokenMigrationFormFull, ITxData, ITxObject, ITxToAddress } from '@types';
import { inputGasPriceToHex, inputValueToHex, toWei } from '@utils';

import { MIGRATION_CONTRACT } from './config';

export const createApproveTx = (
  payload: ITokenMigrationFormFull
): Omit<ITxObject, 'nonce' | 'gasLimit'> =>
  formatApproveTx({
    contractAddress: payload.asset.contractAddress as ITxToAddress,
    baseTokenAmount: toWei(payload.amount, DEFAULT_ASSET_DECIMAL),
    fromAddress: payload.account.address,
    spenderAddress: MIGRATION_CONTRACT,
    chainId: DEFAULT_NETWORK_CHAINID,
    hexGasPrice: inputGasPriceToHex(payload.gasPrice)
  });

export const createMigrationTx = (
  payload: ITokenMigrationFormFull
): Omit<ITxObject, 'nonce' | 'gasLimit'> => {
  const data = AaveMigrator.migrateFromLEND.encodeInput({
    amount: toWei(payload.amount, DEFAULT_ASSET_DECIMAL)
  });
  return {
    from: payload.account.address,
    to: MIGRATION_CONTRACT,
    value: inputValueToHex('0'),
    data: data as ITxData,
    gasPrice: inputGasPriceToHex(payload.gasPrice),
    chainId: DEFAULT_NETWORK_CHAINID
  };
};
