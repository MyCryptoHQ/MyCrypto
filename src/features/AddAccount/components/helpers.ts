import { DerivationPath as DPath } from '@mycrypto/wallets';

import { bigify, hasBalance } from '@utils';

import { TableAccountDisplay } from './HDWTable';

export const sortAccounts = (
  accounts: TableAccountDisplay[],
  displayEmptyAddresses: boolean,
  selectedDPath: DPath
) => {
  const selectedAccounts = accounts
    .filter(({ isSelected, balance }) => (isSelected && balance) || hasBalance(balance))
    .sort((a, b) => (bigify(a.balance!).isGreaterThan(b.balance!) ? -1 : 1));
  const deselectedAccounts = accounts
    .filter(
      ({ isSelected, pathItem: { baseDPath } }) =>
        !isSelected && baseDPath.path === selectedDPath.path
    )
    .sort((a, b) => (a.pathItem.index < b.pathItem.index ? -1 : 1));
  return displayEmptyAddresses ? [...selectedAccounts, ...deselectedAccounts] : selectedAccounts;
};

export const calculateDPathOffset = (accounts: TableAccountDisplay[], selectedDPath: DPath) =>
  Math.max(
    ...accounts
      .filter((acc) => acc.pathItem.baseDPath.path === selectedDPath.path)
      .map(({ pathItem: { index } }) => index)
  ) + 1; // Start scanning from the next index
