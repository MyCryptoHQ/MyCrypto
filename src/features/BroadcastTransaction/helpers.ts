export const hasCamera = async (): Promise<boolean> => {
  if (!navigator.mediaDevices) {
    return false;
  }

  const devices = await navigator.mediaDevices.enumerateDevices();
  return !!devices.find((device) => device.kind === 'videoinput');
};
