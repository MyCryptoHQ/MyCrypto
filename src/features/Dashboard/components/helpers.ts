import approval from '@assets/images/transactions/approval.svg';
import contractDeploy from '@assets/images/transactions/contract-deploy.svg';
import contractInteract from '@assets/images/transactions/contract-interact.svg';
import defizap from '@assets/images/transactions/defizap.svg';
import inbound from '@assets/images/transactions/inbound.svg';
import membershipPurchase from '@assets/images/transactions/membership-purchase.svg';
import outbound from '@assets/images/transactions/outbound.svg';
import swap from '@assets/images/transactions/swap.svg';
import transfer from '@assets/images/transactions/transfer.svg';
import { translateRaw } from '@translations';
import { Asset, ITxTypeMeta } from '@types';

import { ITxHistoryType } from '../types';
import { ITxTypeConfigObj } from './RecentTransactionList';

export const constructTxTypeConfig = ({ type, protocol }: ITxTypeMeta): ITxTypeConfigObj => ({
  label: (asset: Asset) => {
    switch (type) {
      default:
        return protocol
          ? translateRaw('RECENT_TX_LIST_PLATFORM_INTERACTION', {
              $platform: translateRaw(protocol, { $ticker: asset.ticker }),
              $action: translateRaw(`PLATFORM_${type}`, { $ticker: asset.ticker })
            })
          : translateRaw(`PLATFORM_${type}`, { $ticker: asset.ticker });
      case 'GENERIC_CONTRACT_CALL' as ITxHistoryType:
      case ITxHistoryType.CONTRACT_INTERACT:
        return translateRaw('RECENT_TX_LIST_LABEL_CONTRACT_INTERACT', {
          $ticker: asset.ticker || translateRaw('UNKNOWN')
        });
      case ITxHistoryType.INBOUND:
        return translateRaw('RECENT_TX_LIST_LABEL_RECEIVED', {
          $ticker: asset.ticker || translateRaw('UNKNOWN')
        });
      case ITxHistoryType.OUTBOUND:
        return translateRaw('RECENT_TX_LIST_LABEL_SENT', {
          $ticker: asset.ticker || translateRaw('UNKNOWN')
        });
      case ITxHistoryType.TRANSFER:
        return translateRaw('RECENT_TX_LIST_LABEL_TRANSFERRED', {
          $ticker: asset.ticker || translateRaw('UNKNOWN')
        });
      case ITxHistoryType.REP_TOKEN_MIGRATION:
      case ITxHistoryType.GOLEM_TOKEN_MIGRATION:
      case ITxHistoryType.ANT_TOKEN_MIGRATION:
        return protocol
          ? translateRaw('RECENT_TX_LIST_PLATFORM_INTERACTION', {
              $platform: translateRaw(protocol, { $ticker: asset.ticker }),
              $action: translateRaw(`PLATFORM_MIGRATION`, { $ticker: asset.ticker })
            })
          : translateRaw(`PLATFORM_MIGRATION`, { $ticker: asset.ticker });
      case ITxHistoryType.PURCHASE_MEMBERSHIP:
        return translateRaw('PLATFORM_MEMBERSHIP_PURCHASED');
    }
  },
  icon: () => {
    switch (type) {
      case 'DEPOSIT':
      case 'WITHDRAW':
      case 'DEPOSIT_TOKEN':
      case 'ROUTER_TO':
      case 'BORROW':
      case 'REPAY':
      case 'MINT':
      case 'NAME_REGISTERED':
      case 'NAME_RENEWED':
      case 'CANCEL_ORDER':
      case 'GENERIC_CONTRACT_CALL':
      case 'CONTRACT_INTERACT':
      case 'NFT_MINT':
        return contractInteract;
      case 'EXCHANGE':
      case 'SWAP':
      case 'BRIDGE_IN':
      case 'BRIDGE_OUT':
      case 'NFT_EXCHANGE':
        return swap;
      case 'APPROVAL':
      case 'APPROVE':
      case 'NFT_APPROVE':
        return approval;
      case 'FAUCET':
      case 'MINING_PAYOUT':
      case 'CLAIM':
      case 'INBOUND':
        return inbound;
      case 'OUTBOUND':
        return outbound;
      case 'PURCHASE_MEMBERSHIP':
        return membershipPurchase;
      case 'DEFIZAP':
        return defizap;
      case 'CONTRACT_DEPLOY':
        return contractDeploy;
      default:
        return transfer;
    }
  }
});
