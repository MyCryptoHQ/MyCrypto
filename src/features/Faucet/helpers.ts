import axios from 'axios';

import { FAUCET_API } from '@config';
import { StoreAccount } from '@types';

const api = axios.create({
  baseURL: FAUCET_API,
  validateStatus: () => true
});

const throwFailure = (message?: string) => {
  throw new Error(message);
};

const throwApiFailure = () => throwFailure('API_FAILURE');

export const possibleSolution = (solution: string) => {
  return /^[a-zA-Z0-9]{4}$/.test(solution);
};

export const requestChallenge = async (recipientAddress: StoreAccount) => {
  const network = recipientAddress.network.name.toLowerCase();
  const address = recipientAddress.address;
  const result = await api.get(`/challenge/${network}/${address}`).catch(throwApiFailure);

  if (result.status !== 200) {
    throwApiFailure();
  } else if (!result.data.success) {
    throwFailure(result.data.message);
  } else {
    return result.data.result;
  }
};

export const solveChallenge = async (id: string, solution: string) => {
  const result = await api.get(`/solve/${id}/${solution}`).catch(throwApiFailure);

  if (result.status !== 200) {
    throwApiFailure();
  } else if (!result.data.success) {
    throwFailure(result.data.message);
  } else {
    return result.data.result;
  }
};

export const regenerateChallenge = async (id: string) => {
  const result = await api.get(`/regenerate/${id}`).catch(throwApiFailure);

  if (result.status !== 200) {
    throwApiFailure();
  } else if (!result.data.success) {
    throwFailure(result.data.message);
  } else {
    return result.data.result;
  }
};
