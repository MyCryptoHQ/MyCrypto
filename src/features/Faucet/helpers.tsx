import axios from 'axios';
import { bigNumberify, formatEther } from 'ethers/utils';

import { FAUCET_API } from '@config';
import { getBaseAssetByNetwork } from '@services';
import {
  Asset,
  Contact,
  ExtendedAsset,
  ExtendedContact,
  ITxConfig,
  ITxNonce,
  ITxReceipt,
  ITxStatus,
  ITxType,
  ITxValue,
  Network,
  NetworkId,
  StoreAccount,
  TAddress
} from '@types';

import { ITxFaucetResult } from './types';

const api = axios.create({
  baseURL: FAUCET_API,
  validateStatus: () => true
});

export const possibleSolution = (solution: string) => {
  return /^[a-zA-Z0-9]{4}$/.test(solution);
};

export const requestChallenge = async (recipientAddress: StoreAccount) => {
  const network = recipientAddress.network.name.toLowerCase();
  const address = recipientAddress.address;
  const result = await api.get(`/challenge/${network}/${address}`);

  if (result.status !== 200) {
    throw new Error('API_FAILURE');
  } else if (!result.data.success) {
    throw new Error(result.data.message);
  } else {
    return result.data.result;
  }
};

export const solveChallenge = async (id: string, solution: string) => {
  const result = await api.get(`/solve/${id}/${solution}`);

  if (result.status !== 200) {
    throw new Error('API_FAILURE');
  } else if (!result.data.success) {
    throw new Error(result.data.message);
  } else {
    return result.data.result;
  }
};

export const regenerateChallenge = async (id: string) => {
  const result = await api.get(`/regenerate/${id}`);

  if (result.status !== 200) {
    throw new Error('API_FAILURE');
  } else if (!result.data.success) {
    throw new Error(result.data.message);
  } else {
    return result.data.result;
  }
};

const getNetworkByLowercaseId = (id: string, networks: Network[] = []): Network => {
  const capitalizedNetworkId = (id.charAt(0).toUpperCase() + id.slice(1)) as NetworkId;
  return networks.find((network: Network) => network.id === capitalizedNetworkId) as Network;
};

const getFaucetContact = (
  txResult: ITxFaucetResult,
  networks: Network[],
  getContactByAddressAndNetworkId: (
    address: TAddress,
    networkId: NetworkId
  ) => ExtendedContact | undefined,
  createContact: (item: Contact) => void
): ExtendedContact => {
  const network = getNetworkByLowercaseId(txResult.network, networks);
  const existingContact = getContactByAddressAndNetworkId(txResult.from, network.id);

  if (existingContact) {
    return existingContact;
  } else {
    createContact({
      address: txResult.from,
      label: 'MyCrypto Faucet',
      network: network.id,
      notes: ''
    });

    const newContact: ExtendedContact = getContactByAddressAndNetworkId(txResult.from, network.id)!;
    return newContact;
  }
};

export const makeTxConfig = (
  txResult: ITxFaucetResult,
  networks: Network[],
  assets: ExtendedAsset[],
  getContactByAddressAndNetworkId: (
    address: TAddress,
    networkId: NetworkId
  ) => ExtendedContact | undefined,
  createContact: (item: Contact) => void
): ITxConfig => {
  const network = getNetworkByLowercaseId(txResult.network, networks);
  const baseAsset: Asset = getBaseAssetByNetwork({
    network,
    assets
  })!;

  /*
   * ITxConfig.senderAccount uses type StoreAccount, but in this case the user is the recipient and the faucet is the sender.
   * getFaucetContact() returns ExtendedContact, which is the closest we can get.
   * The result is casted to `any` to make it compatible with ITxConfig.
   */
  const senderContact: any = getFaucetContact(
    txResult,
    networks,
    getContactByAddressAndNetworkId,
    createContact
  );

  return {
    rawTransaction: {
      to: txResult.to,
      value: txResult.value as ITxValue,
      gasLimit: txResult.gasLimit,
      data: txResult.data,
      gasPrice: txResult.gasPrice,
      nonce: txResult.nonce.toString() as ITxNonce,
      chainId: txResult.chainId,
      from: txResult.from
    },
    amount: formatEther(txResult.value),
    receiverAddress: txResult.to,
    senderAccount: senderContact,
    from: txResult.from,
    asset: baseAsset,
    baseAsset,
    network,
    gasPrice: txResult.gasPrice,
    gasLimit: txResult.gasLimit,
    nonce: txResult.nonce.toString() as ITxNonce,
    data: txResult.data,
    value: txResult.value
  };
};

export const makeTxReceipt = (
  txResult: ITxFaucetResult,
  networks: Network[],
  assets: ExtendedAsset[]
): ITxReceipt => {
  const network = getNetworkByLowercaseId(txResult.network, networks);
  const baseAsset: Asset = getBaseAssetByNetwork({
    network,
    assets
  })!;

  return {
    asset: baseAsset,
    baseAsset,
    txType: ITxType.FAUCET,
    status: ITxStatus.PENDING,
    receiverAddress: txResult.to,
    amount: formatEther(txResult.value),
    data: txResult.data,
    gasPrice: bigNumberify(txResult.gasPrice),
    gasLimit: bigNumberify(txResult.gasLimit),
    to: txResult.to,
    from: txResult.from,
    value: bigNumberify(txResult.value),
    nonce: txResult.nonce.toString(),
    hash: txResult.hash
  };
};
