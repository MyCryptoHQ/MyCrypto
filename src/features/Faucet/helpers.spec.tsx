import { fAssets, fNetworks } from '@fixtures';

import { makeTxConfig, makeTxReceipt, possibleSolution } from './helpers';
import { ITxFaucetResult } from './types';

const exampleTXResult = {
  chainId: 3,
  data: '0x',
  from: '0xa500B2427458D12Ef70dd7b1E031ef99d1cc09f7',
  gasLimit: '21000',
  gasPrice: '1000000000',
  hash: '0x5049c0847681402db2c303847f2f66ac7f3a6caf63119b676374d5781b8d11e9',
  network: 'ropsten',
  nonce: 39,
  to: '0x0000000000000000000000000000000000000000',
  value: '1'
} as ITxFaucetResult;

describe('Faucet helpers', () => {
  describe('Captcha solution regex', () => {
    test('AbC0 should be a valid solution', () => {
      expect(possibleSolution('AbC0')).toBe(true);
    });
    test('AbC01 should not be a valid solution', () => {
      expect(possibleSolution('AbC01')).toBe(false);
    });
  });

  describe('makeTxConfig', () => {
    const getContactByAddressAndNetworkId = jest.fn();
    const createContact = jest.fn();
    test('matches snapshot', async () => {
      expect(
        makeTxConfig(
          exampleTXResult,
          fNetworks,
          fAssets,
          getContactByAddressAndNetworkId,
          createContact
        )
      ).toMatchSnapshot();
    });

    test('createContact was called', () => {
      expect(createContact).toHaveBeenCalledTimes(1);
      expect(createContact).toHaveBeenCalledWith({
        address: '0xa500B2427458D12Ef70dd7b1E031ef99d1cc09f7',
        label: 'MyCrypto Faucet',
        network: 'Ropsten',
        notes: ''
      });
    });
  });

  describe('makeTxReceipt', () => {
    test('matches snapshot', async () => {
      expect(makeTxReceipt(exampleTXResult, fNetworks, fAssets)).toMatchSnapshot();
    });
  });
});
