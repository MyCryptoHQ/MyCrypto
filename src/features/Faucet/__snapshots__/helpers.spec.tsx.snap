// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Faucet helpers makeTxConfig matches snapshot 1`] = `
Object {
  "amount": "0.000000000000000001",
  "asset": Object {
    "decimal": 18,
    "isCustom": false,
    "name": "Ropsten",
    "networkId": "Ropsten",
    "ticker": "RopstenETH",
    "type": "base",
    "uuid": "77de68da-ecd8-53ba-bbb5-8edb1c8e14d7",
  },
  "baseAsset": Object {
    "decimal": 18,
    "isCustom": false,
    "name": "Ropsten",
    "networkId": "Ropsten",
    "ticker": "RopstenETH",
    "type": "base",
    "uuid": "77de68da-ecd8-53ba-bbb5-8edb1c8e14d7",
  },
  "data": "0x",
  "from": "0xa500B2427458D12Ef70dd7b1E031ef99d1cc09f7",
  "gasLimit": "21000",
  "gasPrice": "1000000000",
  "network": Object {
    "baseAsset": "77de68da-ecd8-53ba-bbb5-8edb1c8e14d7",
    "baseUnit": "RopstenETH",
    "blockExplorer": Object {
      "addressPath": "address",
      "blockPath": "block",
      "name": "Etherscan",
      "origin": "https://ropsten.etherscan.io",
      "tokenPath": "token",
      "txPath": "tx",
    },
    "chainId": 3,
    "color": "#adc101",
    "dPaths": Object {
      "LEDGER_NANO_S": Object {
        "label": "Ledger (ETH)",
        "value": "m/44'/60'/0'",
      },
      "TREZOR": Object {
        "label": "Testnet (ETH)",
        "value": "m/44'/1'/0'/0",
      },
      "default": Object {
        "label": "Testnet (ETH)",
        "value": "m/44'/1'/0'/0",
      },
    },
    "gasPriceSettings": Object {
      "initial": 4,
      "max": 40,
      "min": 0.1,
    },
    "id": "Ropsten",
    "isCustom": false,
    "isTestnet": true,
    "name": "Ropsten",
    "nodes": Array [
      Object {
        "name": "ropsten_infura",
        "service": "Infura",
        "type": "infura",
        "url": "https://ropsten.infura.io/v3/c02fff6b5daa434d8422b8ece54c7286",
      },
    ],
  },
  "nonce": "39",
  "rawTransaction": Object {
    "chainId": 3,
    "data": "0x",
    "from": "0xa500B2427458D12Ef70dd7b1E031ef99d1cc09f7",
    "gasLimit": "21000",
    "gasPrice": "1000000000",
    "nonce": "39",
    "to": "0x0000000000000000000000000000000000000000",
    "value": "1",
  },
  "receiverAddress": "0x0000000000000000000000000000000000000000",
  "senderAccount": undefined,
  "value": "1",
}
`;

exports[`Faucet helpers makeTxReceipt matches snapshot 1`] = `
Object {
  "amount": "0.000000000000000001",
  "asset": Object {
    "decimal": 18,
    "isCustom": false,
    "name": "Ropsten",
    "networkId": "Ropsten",
    "ticker": "RopstenETH",
    "type": "base",
    "uuid": "77de68da-ecd8-53ba-bbb5-8edb1c8e14d7",
  },
  "baseAsset": Object {
    "decimal": 18,
    "isCustom": false,
    "name": "Ropsten",
    "networkId": "Ropsten",
    "ticker": "RopstenETH",
    "type": "base",
    "uuid": "77de68da-ecd8-53ba-bbb5-8edb1c8e14d7",
  },
  "data": "0x",
  "from": "0xa500B2427458D12Ef70dd7b1E031ef99d1cc09f7",
  "gasLimit": BigNumber {
    "_hex": "0x5208",
  },
  "gasPrice": BigNumber {
    "_hex": "0x3b9aca00",
  },
  "hash": "0x5049c0847681402db2c303847f2f66ac7f3a6caf63119b676374d5781b8d11e9",
  "nonce": "39",
  "receiverAddress": "0x0000000000000000000000000000000000000000",
  "status": "PENDING",
  "to": "0x0000000000000000000000000000000000000000",
  "txType": "FAUCET",
  "value": BigNumber {
    "_hex": "0x01",
  },
}
`;
